src/main/java/net/dries007/tfc/common/blockentities/BloomeryBlockEntity.java

```diff
@@ -0,0 +1,443 @@
+                if (bloomery.cachedRecipe == null && state.getValue(BloomeryBlock.LIT))
+                {
+                    bloomery.dumpItems();
+                    state = state.setValue(BloomeryBlock.LIT, false);
+                    level.setBlockAndUpdate(bloomery.worldPosition, state);
+                }
+            }
+            if (state.getValue(BloomeryBlock.LIT) && bloomery.getRemainingTicks() <= 0)
+            {
+                if (bloomery.cachedRecipe != null)
+                {
+                    ItemStack result = bloomery.cachedRecipe.getResult(bloomery.getTotalInput());
+                    level.setBlockAndUpdate(bloomery.getInternalBlock(), TFCBlocks.BLOOM.get().defaultBlockState().setValue(BloomBlock.LAYERS, Math.min(result.getCount(), 8)));
+
+                    level.getBlockEntity(bloomery.getInternalBlock(), TFCBlockEntities.BLOOM.get()).ifPresent(bloom -> {
+                        bloom.setBloom(bloomery.cachedRecipe.getResult(bloomery.getTotalInput()));
+                    });
+                }
+
+                bloomery.inputStacks.clear();
+                bloomery.catalystStacks.clear();
+                bloomery.cachedRecipe = null; // Clear recipe
+
+                bloomery.updateMoltenBlock(false);
+                state = state.setValue(BloomeryBlock.LIT, false);
+                level.setBlockAndUpdate(pos, state);
+            }
+
+            // Update multiblock status
+            int newMaxItems = BloomeryBlock.getChimneyLevels(level, bloomery.getInternalBlock()) * 8;
+            Direction direction = state.getValue(BloomeryBlock.FACING);
+
+            //todo: abort process/dump items if !isFormed?
+            if (!BloomeryBlock.isFormed(level, bloomery.getInternalBlock(), direction))
+            {
+                newMaxItems = 0;
+            }
+
+            bloomery.maxInput = newMaxItems;
+            // since amount of catalyst no longer strictly equals number of iron ores/input, maxCatalyst is determined by amount needed to satisfy current amount of input in the bloomery (maxed at 128 just for sanity's sake)
+            if (bloomery.cachedRecipe == null)
+            {
+                bloomery.maxCatalyst = 0;
+            }
+            else
+            {
+                int maxCatalyst = bloomery.getTotalInput().getAmount() / bloomery.cachedRecipe.getInputFluid().getAmount() * bloomery.cachedRecipe.getCatalyst().getCount();
+                bloomery.maxCatalyst = Math.min(maxCatalyst, 128);
+            }
+
+            boolean turnOff = false;
+            while (bloomery.maxInput < bloomery.inputStacks.size())
+            {
+                turnOff = true;
+                // Structure lost one or more chimney levels
+                Helpers.spawnItem(level, bloomery.worldPosition, bloomery.inputStacks.get(0));
+                bloomery.inputStacks.remove(0);
+                bloomery.markForSync();
```

Comment:
Manipulating item stacks like this is a little scary to me. Please copy the item before you spawn it just in case